v2/apiserver/internal/core/mongodb/workers_store.go

```diff
@@ -54,3 +57,95 @@ func (w *workersStore) UpdateStatus(
 	}
 	return nil
 }
+
+func (w *workersStore) Timeout(
+	ctx context.Context,
+	eventID string,
+) error {
+	timedOutTime := time.Now().UTC()
+
+	res, err := w.collection.UpdateOne(
+		ctx,
+		bson.M{
+			"id":                  eventID,
+			"worker.status.phase": core.WorkerPhasePending,
+			"deleted": bson.M{
+				"$exists": false, // Don't grab logically deleted events
+			},
+		},
+		bson.M{
+			"$set": bson.M{
+				"canceled":            timedOutTime,
+				"worker.status.phase": core.WorkerPhaseTimedOut,
+			},
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"error updating status of event %q worker",
+			eventID,
+		)
+	}
+	if res.MatchedCount == 1 {
+		return nil
+	}
```

Comment:
I don't think a pending worker could (or should) timeout, right? The clock only starts ticking after the worker's pod has been created, right? Or am I overlooking something?

Reply:
Oh, yes, if a worker with phase WorkerPhasePending is guaranteed to still be waiting on a pod, we should remove this.  I wasn't sure if there'd be cases where a worker would have a pod during this phase.

Reply:
>  I wasn't sure if there'd be cases where a worker would have a pod during this phase.

Shouldn't be possible because we move workers to a `STARTING` state before trying to create them in the substrate.